### PR TITLE
Fix RepairRun timestamps, honouring different run states

### PR DESCRIPTION
**Pull request by michaelsembwever**
ref:
 - https://github.com/thelastpickle/cassandra-reaper/issues/603

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration008.java
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration008.java]:
```suggestion
   *  delegates to FixRepairSegmentTimestamps as that does everything already.
```

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration011.java
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration011.java]:
```suggestion
   *  delegates to FixRepairSegmentTimestamps as that does everything already.
```

File: src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration012.java
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/java/io/cassandrareaper/storage/cassandra/Migration012.java]:
```suggestion
   *  delegates to FixRepairSegmentTimestamps as that does everything already.
```

**Review by SesquipedalianDefenestrator — CHANGES_REQUESTED** [review]
A few nitpicky comment updates to match the code updates

**Comment by SesquipedalianDefenestrator** [comment]
For the failing test, I suspect the postgres UPDATE statements need = CAST(NULL as timestamp) or possibly = CAST(NULL as TIMESTAMP WITH TIME ZONE)

**Comment by SesquipedalianDefenestrator** [comment]
The Cassandra fix logic makes sense to me, but I'm having a bit of trouble with both instances of `V12_0_0__fix_repair_run_timestamps.sql`.  I don't see where it's setting `end_date` to anything but NULL. Am I missing something?

**Comment by michaelsembwever** [comment]
> The Cassandra fix logic makes sense to me, but I'm having a bit of trouble with both instances of `V12_0_0__fix_repair_run_timestamps.sql`. I don't see where it's setting `end_date` to anything but NULL. Am I missing something?

I'm not sure I follow the question.
Both sql scripts only ever set `end_date` to `NULL`.

This is to meet the precondition here:  https://github.com/thelastpickle/cassandra-reaper/pull/651/files#diff-206e460c603d749f438e724e1023f236R303

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state = 'NOT_STARTED';
```

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'ERROR';
```

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'DONE';
```

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'ABORTED';
```

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'PAUSED';
```

File: src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/h2/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state = 'PAUSED';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state = 'NOT_STARTED';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'ERROR';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'DONE';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'ABORTED';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state != 'PAUSED';
```

File: src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql
Comment (SesquipedalianDefenestrator) [inline src/server/src/main/resources/db/postgres/V12_0_0__fix_repair_run_timestamps.sql]:
```suggestion
AND state = 'ABORTED';
```

**Review by SesquipedalianDefenestrator — CHANGES_REQUESTED** [review]
Sorry, I was wrong on the cause of the type mismatch on the SQL migrations (or there's multiple).
H2 and Postgres both use a string representation for State.

**Comment by SesquipedalianDefenestrator** [comment]
What I mean is, the original error that prompted this change is that RepairRun.endTime being set to null causes an error and a failure to purge old entries.  `FixRepairRunTimestamps.java` now sets endTime = startTime for ABORTED runs, which prevents that problem for the Cassandra back-end.

For H2 and Postgres backend migrations, though, I can't find an equivalent that's setting endTime = startTime for ABORTED runs.  Don't we need to do something like:
```
UPDATE repair_run
SET end_time = start_time
WHERE end_time is NULL
AND status = 'ABORTED';
```
(isTerminated() looks to be true for state 'ABORTED', so end_time is allowed and required for that state from what I can tell, but won't be set on older aborted runs)

**Comment by michaelsembwever** [comment]
> What I mean is, the original error that prompted this change is that RepairRun.endTime being set to null causes an error and a failure to purge old entries. FixRepairRunTimestamps.java now sets endTime = startTime for ABORTED runs, which prevents that problem for the Cassandra back-end.

my boolean logic was whack. have replaced the "terminated" `SELECT` statements with a `WHERE state NOT IN (..)`

> For H2 and Postgres backend migrations, though, I can't find an equivalent that's setting endTime = startTime for ABORTED runs. Don't we need to do something like:

Indeed! Completely went by me, thanks for picking up on it. Fixed (see both `V12..cql` scripts).

**Review by SesquipedalianDefenestrator — APPROVED** [review]
Revamped logic makes total sense to me now.
Failed tests don't look as concerning, anyways.
One of them was a "git clone" failure, so I'll chalk that up to test infra
One of the test failures I don't understand, but smells a bit like a flaky test rather than a true failure.
The flapping reaper test looks like maybe a credential mismatch?

You're probably more familiar with the project's test infra state than I am, but the failures all look likely to be unrelated to these changes from what I can tell.

**Comment by michaelsembwever** [comment]
> One of the test failures I don't understand, but smells a bit like a flaky test rather than a true failure.

Yes it looks that way!

I am taking your last comment as a reviewer `+1` , hope this is correct?
Once I get the tests passing I will merge.

**Comment by SesquipedalianDefenestrator** [comment]
Yep, +1.  I thought I'd had "Approve" selected, but either I didn't or it didn't count it as a real approval because I don't have write access.

**Comment by michaelsembwever** [comment]
Thanks. Yes you did approve the changes :-)

**Pull request merged**
